File: sdk/java/ScyKernel.java
import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.io.IOException;
import java.io.FileOutputStream;
import java.util.Arrays;
import java.nio.ByteBuffer;
import java.util.zip.*;

public class ScyKernel {
    private String password;
    private String filePath;
    private int hVal;
    private final int canvasSize = 4000;
    private byte[] dbBuffer;

    public ScyKernel(String pwd, String path) {
        this.password = pwd;
        this.filePath = path;
        this.hVal = getHVal(pwd);
        this.dbBuffer = new byte[canvasSize * canvasSize * 3];
    }

    private int getHVal(String pwd) {
        int hash = 7;
        for (int i = 0; i < pwd.length(); i++) {
            hash = hash * 31 + pwd.charAt(i);
        }
        return (int) (((double) (hash & 0xFFFFFFFFL) / 4294967296.0) * 16000000);
    }

    // Deterministic FNV-1a Hash + Alphabet Salt for Cross-Language Parity
    private int deriveIndex(String key, String password) {
        long hash = 0x811c9dc5L & 0xffffffffL; // FNV offset basis (uint32)
        long prime = 0x01000193L;              // FNV prime
        long alphaSalt = 0;
        for (byte b : password.getBytes()) {
            hash ^= (b & 0xFF); // Ensure unsigned byte treatment
            hash = (hash * prime) & 0xFFFFFFFFL; // Mask to 32-bit
        }
        String lowerKey = key.toLowerCase();
        for (int i = 0; i < key.length(); i++) {
            char c = key.charAt(i);
            // FNV-1a Math (keep it 32-bit unsigned logic)
            hash ^= (key.charAt(i) & 0xff);
            hash = (hash * prime) & 0xffffffffL;
            // Alphabet Salt Math (A=1, B=2...)
            if (Character.isLetter(c)) {
                alphaSalt += (lowerKey.charAt(i) - 'a' + 1);
            }
        }
        long finalHash = (hash + alphaSalt) & 0xFFFFFFFFL;
        double normalized = ((double) finalHash / 4294967296.0) * 16000000.0;
        return (int) Math.floor(normalized);
    }

    private void d2xy(int n, int d, int[] coords) {
        int x = 0, y = 0, t = d;
        for (int s = 1; s < n; s *= 2) {
            int rx = 1 & (t / 2);
            int ry = 1 & (t ^ rx);
            rot(s, x, y, rx, ry, coords);
            x = coords[0];
            y = coords[1];
            x += s * rx;
            y += s * ry;
            t /= 4;
        }
        coords[0] = x;
        coords[1] = y;
    }

    private void rot(int n, int x, int y, int rx, int ry, int[] coords) {
        if (ry == 0) {
            if (rx == 1) {
                x = n - 1 - x;
                y = n - 1 - y;
            }
            int temp = x;
            x = y;
            y = temp;
        }
        coords[0] = x;
        coords[1] = y;
    }

    public byte cryptByte(byte data, String password, int position) {
        if (password == null || password.isEmpty()) return data;
        int charIndex = position % password.length();
        int salt = password.charAt(charIndex) ^ position;
        int result = (data & 0xFF) ^ (salt & 0xFF);
        return (byte) result;
    }

    public int compute_crc(byte[] buf, int len) {
        int crc = 0xFFFFFFFF;
        for (int i = 0; i < len; i++) {
            crc ^= (buf[i] & 0xFF);
            for (int j = 0; j < 8; j++) {
                if ((crc & 1) != 0) {
                    crc = (crc >>> 1) ^ 0xEDB88320;
                } else {
                    crc = (crc >>> 1);
                }
            }
        }
        return ~crc;
    }

    public void write32(OutputStream out, int val) throws IOException {
        byte[] b = new byte[] {
            (byte)(val >>> 24),
            (byte)(val >>> 16),
            (byte)(val >>> 8),
            (byte)val
        };
        out.write(b);
    }

    private void writePNGChunk(OutputStream out, String type, byte[] data) throws IOException {
        write32(out, data.length);
        byte[] typeBytes = type.getBytes();
        byte[] crcTarget = new byte[typeBytes.length + data.length];
        System.arraycopy(typeBytes, 0, crcTarget, 0, typeBytes.length);
        System.arraycopy(data, 0, crcTarget, typeBytes.length, data.length);
        out.write(crcTarget);
        write32(out, compute_crc(crcTarget, crcTarget.length));
    }

    public void putToPPM(String key, String value, String password) {
        int index = deriveIndex(key, password);
        int curD = hVal + (index * 1600);
        // Coordinates derived from fractal mapping
        int[] coords = new int[2];
        d2xy(canvasSize, curD, coords);
        int x = coords[0];
        int y = coords[1];
        // RandomAccessFile allows simultaneous read/write with seeking
        try (RandomAccessFile file = new RandomAccessFile(filePath, "rw")) {
            // Header offset (approx P6 PPM)
            long offset = 15 + ((long) y * canvasSize + x) * 3;
            file.seek(offset);
            int i = 0; // Positional tracker for cryptByte
            for (char c : value.toCharArray()) {
                byte[] pixel = new byte[3];
                // Read 3 bytes (the current pixel)
                file.read(pixel);
                // Apply stream cipher before storage
                byte secureChar = cryptByte((byte) c, password, i);
                // Direct assignment to Red channel to match putToPNG logic
                pixel[0] = secureChar;
                // Move back 3 bytes to overwrite the SAME pixel we just read
                file.seek(file.getFilePointer() - 3);
                file.write(pixel);
                i++;
            }
            // Write Null Terminator (0 in Red channel)
            byte[] term = {0, 0, 0};
            file.write(term);
        } catch (IOException e) {
            System.err.println("❌ PPM Write Error: " + e.getMessage());
        }
    }

    public String getFromPPM(String key, String password) {
        int index = deriveIndex(key, password);
        int curD = hVal + (index * 1600);
        int[] coords = new int[2];
        d2xy(canvasSize, curD, coords);
        int x = coords[0];
        int y = coords[1];
        try (RandomAccessFile file = new RandomAccessFile(filePath, "r")) {
            long offset = 15 + ((long) y * canvasSize + x) * 3;
            file.seek(offset);
            StringBuilder result = new StringBuilder();
            int i = 0;
            while (true) {
                byte[] pixel = new byte[3];
                int bytesRead = file.read(pixel);
                // Break on Null Terminator (Red == 0) or EOF
                if (bytesRead == -1 || pixel[0] == 0) break;
                byte scrambled = pixel[0];
                result.append((char) (cryptByte(scrambled, password, i) & 0xFF));
                i++;
            }
            return result.toString();
        } catch (IOException e) {
            return "";
        }
    }

    // PNG sow, harvest functions
    /**
     * Writes a value to the RAM buffer.
     * Note: You MUST call syncPNG(file, "commit") after calling this to save changes.
     */
    public void putToPNG(String key, String value, String keyPassword) {
        // Safety check: ensure buffer is initialized
        if (dbBuffer == null || dbBuffer.length == 0) dbBuffer = new byte[48000000];
        int index = deriveIndex(key, keyPassword);
        int curD = hVal + (index * 1600);
        int[] coords = new int[2];
        d2xy(canvasSize, curD, coords);
        int x = coords[0];
        int y = coords[1];
        // Linear pixel walk in RAM
        for (int i = 0; i < value.length(); i++) {
            // Note: Casting long to handle potential overflow in indexing, though 48MB is safe for int
            int pixelIdx = ((y * canvasSize) + (x + i)) * 3;
            if (pixelIdx + 2 < dbBuffer.length) {
                byte secureChar = cryptByte((byte) value.charAt(i), keyPassword, i);
                // XOR into Red Channel (Bit-perfect data preservation)
                dbBuffer[pixelIdx] = secureChar;
            }
        }
        // Write Null Terminator (0 in Red channel marks the end)
        int termIdx = ((y * canvasSize) + (x + value.length())) * 3;
        if (termIdx + 2 < dbBuffer.length) {
            dbBuffer[termIdx] = 0;
        }
    }

    /**
     * Retrieves a value from the RAM buffer.
     * Note: You SHOULD call syncPNG(file, "load") before this to ensure fresh data.
     */
    public String getFromPNG(String key, String keyPassword) {
        if (dbBuffer == null || dbBuffer.length == 0) return "";
        int index = deriveIndex(key, keyPassword);
        int curD = hVal + (index * 1600);
        int[] coords = new int[2];
        d2xy(canvasSize, curD, coords);
        int x = coords[0];
        int y = coords[1];
        StringBuilder result = new StringBuilder();
        int i = 0;
        while (true) {
            int pixelIdx = ((y * canvasSize) + (x + i)) * 3;
            // Bounds check + Null Terminator check (Red == 0)
            if (pixelIdx + 2 >= dbBuffer.length || dbBuffer[pixelIdx] == 0) {
                break;
            }
            byte scrambled = dbBuffer[pixelIdx];
            result.append((char) (cryptByte(scrambled, keyPassword, i) & 0xFF));
            i++;
        }
        return result.toString();
    }

    // DB sync functions for easier DB handling
    public boolean syncPNG(String filename, String mode) {
        String lowerMode = mode.toLowerCase();
        if (lowerMode.equals("load") && !Files.exists(Paths.get(filename))) {
            System.out.println("⚠️ Database not found. Initializing new compressed store...");
            Arrays.fill(dbBuffer, (byte) 0);
            return syncPNG(filename, "commit");
        }
        if (lowerMode.equals("commit")) {
            try (FileOutputStream fos = new FileOutputStream(filename)) {
                // Prepare Filtered Data
                byte[] filtered = new byte[48004000];
                for (int r = 0; r < 4000; r++) {
                    filtered[r * 12001] = 0; 
                    System.arraycopy(dbBuffer, r * 12000, filtered, (r * 12001) + 1, 12000);
                }
                // Compress
                Deflater deflater = new Deflater();
                deflater.setInput(filtered);
                deflater.finish();
                byte[] compressed = new byte[48004000];
                int compressedSize = deflater.deflate(compressed);
                deflater.end();
                // Write PNG Structure
                fos.write(new byte[]{(byte)0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A});
                byte[] ihdrBody = new byte[]{
                    0,0,(byte)0x0F,(byte)0xA0, // 4000
                    0,0,(byte)0x0F,(byte)0xA0, // 4000
                    8, 2, 0, 0, 0
                };
                writePNGChunk(fos, "IHDR", ihdrBody);
                writePNGChunk(fos, "IDAT", Arrays.copyOf(compressed, compressedSize));
                writePNGChunk(fos, "IEND", new byte[0]);
                System.out.println("✅ PNG Commit Successful: " + filename);
                return true;
            } catch (IOException e) {
                return false;
            }
        } else if (lowerMode.equals("load")) {
            try (DataInputStream dis = new DataInputStream(new FileInputStream(filename))) {
                dis.skipBytes(33); // Skip Sig and IHDR
                int cLen = dis.readInt();
                dis.skipBytes(4); // Skip 'IDAT'
                byte[] cData = new byte[cLen];
                dis.readFully(cData);
                Inflater inflater = new Inflater();
                inflater.setInput(cData);
                byte[] decomp = new byte[48004000];
                inflater.inflate(decomp);
                inflater.end();
                for (int r = 0; r < 4000; r++) {
                    System.arraycopy(decomp, (r * 12001) + 1, dbBuffer, r * 12000, 12000);
                }
                System.out.println("✅ PNG Load Successful: " + filename);
                return true;
            } catch (IOException | DataFormatException e) {
                return false;
            }
        }
        return false;
    }

    public boolean syncPPM(String filename, String mode) {
        String lowerMode = mode.toLowerCase();
        if (lowerMode.equals("load") && !Files.exists(Paths.get(filename))) {
            System.out.println("⚠️ PPM Database not found. Initializing new raw store...");
            Arrays.fill(dbBuffer, (byte) 0);
            return syncPPM(filename, "commit");
        }
        if (lowerMode.equals("commit")) {
            try (FileOutputStream fos = new FileOutputStream(filename)) {
                fos.write(("P6\n4000 4000\n255\n").getBytes());
                fos.write(dbBuffer);
                System.out.println("✅ PPM Commit Successful: " + filename);
                return true;
            } catch (IOException e) {
                return false;
            }
        } else if (lowerMode.equals("load")) {
            try (FileInputStream fis = new FileInputStream(filename)) {
                fis.skip(15);
                fis.read(dbBuffer);
                System.out.println("✅ PPM Load Successful: " + filename);
                return true;
            } catch (IOException e) {
                return false;
            }
        }
        return false;
    }

    // Blank DB creation functions
    public void createPNG_DB(String filename) {
        Arrays.fill(dbBuffer, (byte) 0);
        if (syncPNG(filename, "commit")) {
            System.out.println("✅ PNG initialized and loaded into buffer: " + filename);
        } else {
            System.err.println("❌ Failed to initialize PNG database file.");
        }
    }

    public void createPPM_DB(String dbPath) {
        try (FileOutputStream ofs = new FileOutputStream(dbPath)) {
            // Write the P6 Header (Standard 4000x4000 8-bit RGB)
            String header = "P6\n" + canvasSize + " " + canvasSize + "\n255\n";
            ofs.write(header.getBytes());
            byte[] zeroRow = new byte[canvasSize * 3];
            Arrays.fill(zeroRow, (byte) 0);
            for (int i = 0; i < canvasSize; i++) {
                ofs.write(zeroRow);
            }
            System.out.println("✅ PPM Database Ready (Isolated from RAM): " + dbPath);
        } catch (IOException e) {
            System.err.println("❌ Error: Could not create PPM database - " + e.getMessage());
        }
    }

    // DB conversion functions
    public boolean convertDatabaseFormat(String pngPath, String ppmPath, String targetFormat) {
        // Standardize the flag to lowercase to handle "PNG", "png", "Ppm", etc.
        String lowerTarget = targetFormat.toLowerCase();
        if (lowerTarget.equals("ppm")) {
            // SOURCE: PNG -> TARGET: PPM (Decompress and Expand)
            if (!syncPNG(pngPath, "load")) {
                System.err.println("❌ Failed to load PNG database.");
                return false;
            }
            try (FileOutputStream ppm = new FileOutputStream(ppmPath)) {
                // Write P6 PPM Header (Standard 4000x4000 8-bit)
                String header = "P6\n4000 4000\n255\n";
                ppm.write(header.getBytes());
                // Dump the raw 48MB buffer into the file
                ppm.write(dbBuffer);
                System.out.println("✅ Converted PNG to PPM (48MB Raw Volume)");
                return true;
            } catch (IOException e) {
                System.err.println("❌ IO Error during PPM export: " + e.getMessage());
                return false;
            }
        } else if (lowerTarget.equals("png")) {
            // SOURCE: PPM -> TARGET: PNG (Pack and Compress)
            try (FileInputStream ppm = new FileInputStream(ppmPath)) {
                // Skip the header (Assuming standard 15-byte offset)
                long skipped = ppm.skip(15);
                if (skipped < 15) {
                    System.err.println("❌ Invalid PPM header encountered.");
                    return false;
                }
                // Read the raw 48MB into our RAM buffer
                int totalRead = 0;
                while (totalRead < 48000000) {
                    int read = ppm.read(dbBuffer, totalRead, 48000000 - totalRead);
                    if (read == -1) break; // EOF
                    totalRead += read;
                }
                // Use syncPNG to compress the buffer and save as a PNG
                if (!syncPNG(pngPath, "commit")) {
                    System.err.println("❌ Failed to compress and save PNG database.");
                    return false;
                }
                System.out.println("✅ Converted PPM to PNG");
                return true;
            } catch (IOException e) {
                System.err.println("❌ IO Error during PNG import: " + e.getMessage());
                return false;
            }
        } else {
            System.err.println("❌ Invalid target format. Use 'PNG' or 'PPM'.");
            return false;
        }
    }

    // DB Deletion and cleanup functions
    public boolean deleteDB(String dbPath) throws IOException {
        Path path = Paths.get(dbPath);
        try {
            // Files.deleteIfExists returns true if the file was there and deleted,
            // and false if the file didn't exist.
            return Files.deleteIfExists(path);
        } catch (IOException e) {
            // Log the error or rethrow depending on your SDK's error policy
            System.err.println("Failed to delete database: " + e.getMessage());
            throw e; 
        }
    }

    // DB checking/validating functions
    public long getFileSize(String pathStr) {
        try {
            Path path = Paths.get(pathStr);
            if (Files.exists(path) && Files.isRegularFile(path)) {
                return Files.size(path);
            }
        } catch (IOException e) {
            System.err.println("❌ Filesystem Error: " + e.getMessage());
        }
        return 0L;
    }

    public boolean validateDB(String path) {
        // 4000 * 4000 * 3
        final long rawDataSize = 48000000L; 
        long actual = getFileSize(path);
        if (path.toLowerCase().endsWith(".ppm")) {
            // PPM must have the header + the data
            return actual >= (rawDataSize + 15);
        } else {
            // PNG/Raw must be at least the data size
            return actual >= rawDataSize;
        }
    }

    // Extra functions

}

File: sdk/kotlin/ScyKernel.kt
import java.io.RandomAccessFile
import java.io.File
import kotlin.math.*
import java.util.Locale
import java.nio.file.Files
import java.nio.file.Paths

class ScyKernel(private val password: String, private val filePath: String) {
    private val canvasSize: Int = 4000
    private var hVal: Long = 0
    private var dbBuffer: ByteArray? = null

    init {
        this.hVal = getHVal(password).toLong()
        this.dbBuffer = null
    }

    private fun getHVal(pwd: String): Int {
        var hash: Int;
        hash = 7
        for (c in pwd) {
            hash = hash * 31 + c.code
        }
        var unsignedHash: Long;
        unsignedHash = hash.toLong() and 0xFFFFFFFFL
        var normalized: Double;
        normalized = unsignedHash.toDouble() / 4294967296.0
        var result: Int;
        result = floor(normalized * 16000000.0).toInt()
        return result
    }

    private fun _deriveIndex(key: String, password: String): Int {
        var hash = 0x811c9dc5L
        val prime = 0x01000193L
        var alphaSalt = 0L
        for (c in password) {
            hash = hash xor (c.code.toLong() and 0xFFL)
            hash = (hash * prime) and 0xFFFFFFFFL
        }
        for (c in key) {
            hash = hash xor (c.code.toLong() and 0xFFL)
            hash = (hash * prime) and 0xFFFFFFFFL
            if (c.isLetter()) {
                alphaSalt += (c.lowercaseChar().code - 'a'.code + 1).toLong()
            }
        }
        val finalVal = (hash + alphaSalt) and 0xFFFFFFFFL
        val normalized = (finalVal.toDouble() / 4294967296.0) * 16000000.0
        return kotlin.math.floor(normalized).toInt()
    }

    private fun _rot(n: Int, x: Int, y: Int, rx: Int, ry: Int): Pair<Int, Int> {
        var tx = x
        var ty = y
        if (ry == 0) {
            if (rx == 1) {
                tx = n - 1 - x
                ty = n - 1 - y
            }
            return Pair(ty, tx)
        }
        return Pair(tx, ty)
    }

    private fun _d2xy(n: Int, d: Long): Pair<Int, Int> {
        var x = 0
        var y = 0
        var t = d
        var s = 1
        while (s < n) {
            val rx = (1L and (t / 2)).toInt()
            val ry = (1L and (t xor rx.toLong())).toInt()
            val (nx, ny) = _rot(s, x, y, rx, ry)
            x = nx + s * rx
            y = ny + s * ry
            t /= 4
            s *= 2
        }
        return Pair(x, y)
    }

    /**
     * A bit-perfect, zero-byte overhead encryption layer.
     * XORs the character based on the password's hash and the character's position.
     */
    private fun _cryptByte(c: Char, password: String, position: Int): Char {
        var salt: UInt = 0x811c9dc5u 
        for (pc in password) {
            salt = (salt xor pc.code.toUInt()) * 16777619u 
        }
        var mixed: UInt = salt xor (position.toUInt() * 0xdeadbeefu)
        mixed = mixed xor (mixed shr 16)
        val keyByte = (mixed and 0xFFu).toByte()
        return (c.code xor (keyByte.toInt() and 0xFF)).toChar()
    }

    // Lightweight CRC-32 for PNG Chunk Compliance
    private fun _compute_crc(buf: ByteArray, len: Int): UInt {
        var crc: UInt = 0xFFFFFFFFu
        for (i in 0 until len) {
            crc = crc xor (buf[i].toUInt() and 0xFFu)
            for (j in 0 until 8) {
                // Implementing -(crc & 1) logic via mask
                val mask = if ((crc and 1u) == 1u) 0xEDB88320u else 0u
                crc = (crc shr 1) xor mask
            }
        }
        return crc.inv()
    }

    private fun _write32(out: java.io.OutputStream, val32: UInt) {
        val b = byteArrayOf(
            (val32 shr 24).toByte(),
            (val32 shr 16).toByte(),
            (val32 shr 8).toByte(),
            val32.toByte()
        )
        out.write(b, 0, 4)
    }

    // PPM sow, harvest functions
    /**
    * Writes a value directly to the PPM file on disk.
    */
    fun _putToPPM(key: String, value: String, password: String) {
        val index = _deriveIndex(key, password)
        val curD = hVal + (index * 1600)
        val (x, y) = _d2xy(canvasSize, curD)

        try {
            val file = java.io.RandomAccessFile(filePath, "rw")
            val offset = 15L + (y.toLong() * canvasSize + x) * 3L
            file.seek(offset)
            var i = 0 
            for (c in value) {
                val pixel = ByteArray(3)
                val bytesRead = file.read(pixel)
                if (bytesRead < 3) break
                val secureChar = _cryptByte(c, password, i)
                pixel[0] = secureChar.code.toByte()
                file.seek(file.filePointer - 3)
                file.write(pixel)
                i++
            }
            val term = byteArrayOf(0, 0, 0)
            file.write(term)
            file.close()
        } catch (e: Exception) {}
    }

    fun _getFromPPM(key: String, password: String): String {
        val index = _deriveIndex(key, password)
        val curD = hVal + (index * 1600)
        val (x, y) = _d2xy(canvasSize, curD)
        try {
            val file = java.io.RandomAccessFile(filePath, "r")
            val offset = 15L + (y.toLong() * canvasSize + x) * 3L
            file.seek(offset)
            var result = ""
            var i = 0
            while (true) {
                val pixel = ByteArray(3)
                val bytesRead = file.read(pixel)
                if (bytesRead < 3 || pixel[0].toInt() == 0) break
                val scrambled = (pixel[0].toInt() and 0xFF).toChar()
                result += _cryptByte(scrambled, password, i)
                i++
            }
            file.close()
            return result
        } catch (e: Exception) {
            return ""
        }
    }

    // PNG sow, harvest functions
    /**
    * Writes a value to the RAM buffer.
    * Note: You MUST call syncPNG(file, "commit") after calling this to save changes.
    */
    fun _putToPNG(key: String, value: String, keyPassword: String) {
        if (dbBuffer == null || dbBuffer!!.isEmpty()) {
            dbBuffer = ByteArray(48000000)
        }
        val index = _deriveIndex(key, keyPassword)
        val curD = hVal + (index * 1600)
        val (x, y) = _d2xy(canvasSize, curD)
        for (i in value.indices) {
            val pixelIdx = ((y * canvasSize) + (x + i)) * 3
            if ((pixelIdx + 2) < dbBuffer!!.size) {
                val secureChar = _cryptByte(value[i], keyPassword, i)
                dbBuffer!![pixelIdx] = secureChar.code.toByte()
            }
        }
        val termIdx = ((y * canvasSize) + (x + value.length)) * 3
        if ((termIdx + 2) < dbBuffer!!.size) {
            dbBuffer!![termIdx] = 0.toByte()
        }
    }

    /**
    * Retrieves a value from the RAM buffer.
    * Note: You SHOULD call syncPNG(file, "load") before this to ensure fresh data.
    */
    fun _getFromPNG(key: String, keyPassword: String): String {
        if (dbBuffer == null || dbBuffer!!.isEmpty()) return "" 
        val index = _deriveIndex(key, keyPassword)
        val curD = hVal + (index * 1600)
        val (x, y) = _d2xy(canvasSize, curD)
        var result = ""
        var i = 0
        while (true) {
            val pixelIdx = ((y * canvasSize) + (x + i)) * 3
            if ((pixelIdx + 2) >= dbBuffer!!.size || dbBuffer!![pixelIdx].toInt() == 0) {
                break
            }
            val scrambled = (dbBuffer!![pixelIdx].toInt() and 0xFF).toChar()
            result += _cryptByte(scrambled, keyPassword, i)
            i++
        }
        return result
    }

    // DB sync functions for easier DB handling
    fun _syncPNG(filename: String, modeIn: String): Boolean {
        val mode = modeIn.lowercase()
        if (mode == "load" && !java.io.File(filename).exists()) {
            println("⚠️ Database not found. Initializing new compressed store...")
            dbBuffer = ByteArray(48000000)
            return _syncPNG(filename, "commit")
        }
        if (mode == "commit") {
            val filtered = ByteArray(48004000)
            for (r in 0 until 4000) {
                filtered[r * 12001] = 0
                dbBuffer?.copyInto(filtered, r * 12001 + 1, r * 12000, (r + 1) * 12000)
            }
            val deflater = java.util.zip.Deflater(java.util.zip.Deflater.BEST_COMPRESSION)
            deflater.setInput(filtered)
            deflater.finish()
            val compressedOutput = java.io.ByteArrayOutputStream()
            val compBuf = ByteArray(1024)
            while (!deflater.finished()) {
                val count = deflater.deflate(compBuf)
                compressedOutput.write(compBuf, 0, count)
            }
            val compressed = compressedOutput.toByteArray()
            try {
                val out = java.io.FileOutputStream(filename)
                val sig = byteArrayOf(0x89.toByte(), 0x50.toByte(), 0x4E.toByte(), 0x47.toByte(), 0x0D.toByte(), 0x0A.toByte(), 0x1A.toByte(), 0x0A.toByte())
                out.write(sig)
                val ihdr = byteArrayOf('I'.code.toByte(), 'H'.code.toByte(), 'D'.code.toByte(), 'R'.code.toByte(), 0, 0, 0x0F.toByte(), 0xA0.toByte(), 0, 0, 0x0F.toByte(), 0xA0.toByte(), 8, 2, 0, 0, 0)
                _write32(out, 13u)
                out.write(ihdr)
                _write32(out, _compute_crc(ihdr, 17))
                _write32(out, compressed.size.toUInt())
                val idatTag = "IDAT".toByteArray()
                out.write(idatTag)
                out.write(compressed)
                val crcB = java.io.ByteArrayOutputStream()
                crcB.write(idatTag)
                crcB.write(compressed)
                _write32(out, _compute_crc(crcB.toByteArray(), crcB.size()))
                _write32(out, 0u)
                val iendTag = "IEND".toByteArray()
                out.write(iendTag)
                _write32(out, _compute_crc(iendTag, 4))
                out.close()
                println("✅ PNG Commit Successful: $filename")
                return true
            } catch (e: Exception) {
                return false
            }
        } else if (mode == "load") {
            try {
                val file = java.io.File(filename)
                val fis = java.io.FileInputStream(file)
                val fullData = fis.readBytes()
                fis.close()
                val bis = java.io.ByteArrayInputStream(fullData)
                bis.skip(33)
                val lBuf = ByteArray(4)
                bis.read(lBuf)
                val cLen = ((lBuf[0].toUInt() and 0xFFu) shl 24) or 
                           ((lBuf[1].toUInt() and 0xFFu) shl 16) or 
                           ((lBuf[2].toUInt() and 0xFFu) shl 8) or 
                           (lBuf[3].toUInt() and 0xFFu)
                bis.skip(4)
                val cData = ByteArray(cLen.toInt())
                bis.read(cData)
                val inflater = java.util.zip.Inflater()
                inflater.setInput(cData)
                val decomp = ByteArray(48004000)
                inflater.inflate(decomp)
                inflater.end()
                dbBuffer = ByteArray(48000000)
                for (r in 0 until 4000) {
                    decomp.copyInto(dbBuffer!!, r * 12000, r * 12001 + 1, (r + 1) * 12001)
                }
                //println("✅ PNG Load Successful: $filename")
                return true
            } catch (e: Exception) {
                return false
            }
        } else {
            System.err.println("❌ Invalid Sync Mode. Use 'Commit' or 'Load'.")
            return false
        }
    }

    fun _syncPPM(filename: String, modeIn: String): Boolean {
        val mode = modeIn.lowercase()
        if (mode == "load" && !java.io.File(filename).exists()) {
            println("⚠️ PPM Database not found. Initializing new raw store...")
            dbBuffer = ByteArray(48000000)
            return _syncPPM(filename, "commit")
        }
        if (mode == "commit") {
            try {
                val out = java.io.FileOutputStream(filename)
                out.write("P6\n4000 4000\n255\n".toByteArray())
                dbBuffer?.let { out.write(it) }
                out.close()
                println("✅ PPM Commit Successful: $filename")
                return true
            } catch (e: Exception) {
                return false
            }
        } else if (mode == "load") {
            try {
                val file = java.io.File(filename)
                val fis = java.io.FileInputStream(file)
                fis.skip(15)
                dbBuffer = ByteArray(48000000)
                fis.read(dbBuffer)
                fis.close()
                println("✅ PPM Load Successful: $filename")
                return true
            } catch (e: Exception) {
                return false
            }
        } else {
            System.err.println("❌ Invalid Sync Mode. Use 'Commit' or 'Load'.")
            return false
        }
    }

    // Blank DB creation functions
    fun _createPNG_DB(filename: String) {
        dbBuffer = ByteArray(48000000)
        if (_syncPNG(filename, "commit")) {
            println("✅ PNG initialized and loaded into buffer: $filename")
        } else {
            System.err.println("❌ Failed to initialize PNG database file.")
        }
    }

    fun _createPPM_DB(dbPath: String) {
        try {
            val fos = java.io.FileOutputStream(dbPath)
            // Write the P6 Header (Standard 4000x4000 8-bit RGB)
            val header = "P6\n$canvasSize $canvasSize\n255\n"
            fos.write(header.toByteArray())
            val zeroRow = ByteArray(canvasSize * 3)
            for (i in 0 until canvasSize) {
                fos.write(zeroRow)
            }
            fos.close()
            println("✅ PPM Database Ready (Isolated from RAM): $dbPath")
        } catch (e: Exception) {
            System.err.println("❌ Error: Could not create PPM database")
            return
        }
    }

    // DB conversion functions
    fun _convertDatabaseFormat(pngPath: String, ppmPath: String, targetFormat: String): Boolean {
        // Standardize the flag to lowercase to handle "PNG", "png", "Ppm", etc.
        val format = targetFormat.lowercase()
        if (format == "ppm") { 
            // SOURCE: PNG -> TARGET: PPM (Decompress and Expand)
            if (!_syncPNG(pngPath, "load")) {
                System.err.println("❌ Failed to load PNG database.")
                return false
            }
            try {
                val ppmFile = java.io.File(ppmPath)
                val fos = java.io.FileOutputStream(ppmFile)
                // Write P6 PPM Header (Standard 4000x4000 8-bit)
                fos.write("P6\n4000 4000\n255\n".toByteArray())
                // Dump the raw 48MB buffer into the file
                dbBuffer?.let { fos.write(it) }
                fos.close()
                println("✅ Converted PNG to PPM (48MB Raw Volume)")
                return true
            } catch (e: Exception) {
                return false
            }
        } else if (format == "png") { 
            // SOURCE: PPM -> TARGET: PNG (Pack and Compress)
            try {
                val ppmFile = java.io.File(ppmPath)
                if (!ppmFile.exists()) return false
                val fis = java.io.FileInputStream(ppmFile)
                // Skip the header (Assuming "P6\n4000 4000\n255\n" is 15 bytes)
                fis.skip(15)
                // Read the raw 48MB into our RAM buffer
                val rawDataSize = 48000000
                dbBuffer = ByteArray(rawDataSize)
                fis.read(dbBuffer, 0, rawDataSize)
                fis.close()
                // Use syncPNG to compress and save as a PNG
                if (!_syncPNG(pngPath, "commit")) {
                    System.err.println("❌ Failed to compress and save PNG database.")
                    return false
                }
                println("✅ Converted PPM to PNG")
                return true
            } catch (e: Exception) {
                return false
            }
        } else {
            System.err.println("❌ Invalid target format. Use 'PNG' or 'PPM'.")
            return false
        }
    }

    // DB Deletion and cleanup functions
    fun deleteDB(dbPath: String): Boolean {
        return Files.deleteIfExists(Paths.get(dbPath))
    }

    // DB checking/validating functions
    fun _getFileSize(path: String): Long {
        try {
            val file = java.io.File(path)
            if (file.exists() && file.isFile) {
                return file.length()
            }
        } catch (e: Exception) {
            System.err.println("❌ Filesystem Error: ${e.message}")
        }
        return 0
    }

    fun _validateDB(path: String): Boolean {
        // 4000 * 4000 * 3
        val rawDataSize: Long = 48000000
        val actual = _getFileSize(path)
        if (path.contains(".ppm")) {
            // PPM must have the header + the data
            return actual >= (rawDataSize + 15)
        } else {
            // PNG/Raw must be at least the data size
            return actual >= rawDataSize
        }
    }

}